packages/source-map/lib/sourceMap.ts

```diff
@@ -99,11 +174,192 @@ export class SourceMap<Data = unknown> {
 				}
 			}
 		}
+		this.offsets = offsets
+		this.mappings = mappings
+	}
+
+	* findMatchingMappingIndexes(offset: number, filter: ((data: Data) => boolean) | undefined): Generator<[Mapping<Data>, number]> {
+		if (this.offsets.length === 0) {
+			return;
+		}
+
+		const fromRange = this.key
+
+		const { low: start, high: end } = binarySearch(this.offsets, offset);
+		const skip = new Set<Mapping>();
+
+		for (let i = start; i <= end; i++) {
+			for (const mapping of this.mappings[i]) {
+				if (skip.has(mapping)) {
+					continue;
+				}
+				skip.add(mapping);
+				if (filter && !filter(mapping.data)) {
+					continue;
+				}
+
+				const fromOffsets = mapping[fromRange]
+				const fromLengths = getLengths(mapping, fromRange)
+				let low = 0;
+				let high = fromOffsets.length - 1;
+
+				while (low <= high) {
+					const mid = Math.floor((low + high) / 2);
+					const fromOffset = fromOffsets[mid];
+					const fromLength = fromLengths[mid];
+
+					if (offset >= fromOffset && offset <= fromOffset + fromLength) {
+						yield [mapping, mid]
```

Comment:
The only difference in the algorithm here is that I return the index instead of ready-to-use offset, as I can than match ranges, so one would not need to do zero-length thingies.